control-plane/connect-inject/terminatinggatewayservice_controller.go

```diff
@@ -0,0 +1,440 @@
+		}
+	} else {
+		if containsString(terminatingGatewayService.Finalizers, FinalizerName) {
+			r.Log.Info("TerminatingGatewayService was deleted, deleting from consul", "name", req.Name, "ns", req.Namespace)
+			err := r.deleteService(spec.Service.ServiceName)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			controllerutil.RemoveFinalizer(terminatingGatewayService, FinalizerName)
+			err = r.Update(ctx, terminatingGatewayService)
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, err
+}
+
+func terminatingGatewayACLRole(aclRoleList []*api.ACLRole) (*api.ACLRole, error) {
+	strToFind := "terminating-gateway"
+
+	result := &api.ACLRole{}
+	roleFound := false
+
+	for _, role := range aclRoleList {
+		if strings.Contains(role.Name, strToFind) {
+			result = role
+			roleFound = true
+			break
+		}
+	}
+
+	if !roleFound {
+		return result, errors.New("Terminating Gateway ACL Role not found")
+	}
+	return result, nil
+}
+
+// updateStatus updates the terminatingGatewayService's ReconcileError in the status.
+func (r *TerminatingGatewayServiceController) updateStatus(ctx context.Context, terminatingGatewayService *consulv1alpha1.TerminatingGatewayService) error {
+
+	policyName := ""
+	if r.AclEnabled {
+		policyName = terminatingGatewayService.Spec.Service.ServiceName + "-write-policy"
+	}
+
+	terminatingGatewayService.Status.LastSyncedTime = &metav1.Time{Time: time.Now()}
+	terminatingGatewayService.SetSyncedCondition(corev1.ConditionTrue, "", "")
+
+	terminatingGatewayService.Status.ServiceInfoRef = &consulv1alpha1.ServiceInfoRefStatus{
+		ServiceName: terminatingGatewayService.Spec.Service.ServiceName,
+		PolicyName:  policyName,
+	}
+
+	err := r.Status().Update(ctx, terminatingGatewayService)
+	if err != nil {
+		r.Log.Error(err, "failed to update TerminatingGatewayService status", "name", terminatingGatewayService.Name, terminatingGatewayService.Namespace)
+	}
+	return err
+}
+
+// updateStatusError updates the terminatingGatewayService's ReconcileError in the status.
+func (r *TerminatingGatewayServiceController) updateStatusError(ctx context.Context, terminatingGatewayService *consulv1alpha1.TerminatingGatewayService, reconcileErr error) {
+	terminatingGatewayService.SetSyncedCondition(corev1.ConditionFalse, "Error updating status", reconcileErr.Error())
+
+	err := r.Status().Update(ctx, terminatingGatewayService)
+	if err != nil {
+		r.Log.Error(err, "failed to update TerminatingGatewayService status", "name", terminatingGatewayService.Name, terminatingGatewayService.Namespace)
+	}
+
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *TerminatingGatewayServiceController) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&consulv1alpha1.TerminatingGatewayService{}).
+		Complete(r)
+}
+
+func (r *TerminatingGatewayServiceController) createOrUpdateService(terminatingGatewayService *consulv1alpha1.TerminatingGatewayService, ctx context.Context) error {
```

Comment:
When we add a comment for this function can we mention that it creates or updates a service in Consul? Just to distinguish from kubernetes services.

Reply:
Done